### DISCUSSION TITLE
Detecting bold fonts in a pdf

### DISCUSSION THREAD

**Discussion by theshahshow** [pymupdf/PyMuPDF#3779] [Looking for help] [answered]
To check whether the text inside a span is bold or not, i usually check if 'bold' exists in span['font'] (font name) or not. 

But in the attached pdf [bdh_single.pdf](https://github.com/user-attachments/files/16605871/bdh_single.pdf), all the font's are "SourceSansPro-Regular". When looking at the pdf, you can clearly see that some text (headings) are bold. In such cases, how do i find if some text is bold or not?

For font names, i look at the ouput of this code:

doc = fitz.open("bdh_single.pdf")
page = doc[0]

blocks = page.get_text("dict")["blocks"]
for block in blocks:
    if block["type"] == 0:
        for line in block["lines"]:
            for span in line['spans']:
                 font_name = span['font']

**Comment by JorjMcKie** [comment]
Of course, PyMuPDF has no way of visually determining how a glyph looks like for the human eye: it is dependent on what the font itself tells us about its properties.
If the font says "I am not bold" we are bound to believe it - no matter how fat the characters appear to be.

Some PDF creators use tricks in an effort to achieve text effects **_without_** embedding another font.
These tricks may include things like writing some text twice - with a little offset (to the right).
Or writing characters with a border around the interior, and more of that sort.

**Reply by JorjMcKie** [reply, marked as answer]
Just had another look:
This PDF uses PDF command `2 Tr` which means "Set text rendering mode 2":
![image](https://github.com/user-attachments/assets/3c47f81a-4a9d-4e7a-8962-29fc339862b4)
So if fill and stroke color are the same (e.g. black) then - with a suitable line width - characters written like that **_appear_** bold, while still the same font is used.

**Reply by theshahshow** [reply]
Thanks for the quick reply!
I get your explanation. How do i find the text rendering mode for a span of text (is the text rendering mode set only once? or can it be different throughout the pdf?). Similarly is there a way to find the fill and stroke color of a text?

**Reply by theshahshow** [reply]
Ok, so going through earlier discussions, I found out how to find the fill and stroke, by using the page.get_texttrace(). Just wondering if any  other solution for this exists or not.

**Reply by JorjMcKie** [reply]
You answered most of your questions yourself 👍!
Command `Tr` can be set for every piece of text. Multiple values are possible per page. Value 3 is often used by OCR engines to store invisible (but extractable and searchable) text upon scanned pages.

Only `get_texttrace()` delivers all those text details - it is **_not_** based on a `TextPage` object but built upon more low-level MuPDF code.

**Reply by theshahshow** [reply]
Got it, thanks!
